bcoin wallet pagination test: docker container and init script

pinheadmz opened the pull request:
After checking out this branch you'll need to rebuild the bcoin container because it pulls from a different repo (my PR on Braydon's PR): https://github.com/pinheadmz/bcoin/tree/wallet-pagination-zip

```
docker-compose build bcoin
docker-compose up bcoin
```
...then run bPanel and switch to `_docker` client to access the changes.

The init script:
- Mines 144 * 3 blocks to `primary/default` to activate SegWit on the regtest chain
- Creates 10 wallets (randomly segwit or not) with 3 accounts each
- Disperses all the funds from the initial mining evenly to all accounts
- Randomly creates transactions between the accounts: Sending acct, number of outputs, receiving accts, send amount. You can see these settings in the first few lines of the init script
- Mines batches of txs into blocks

Also - and this part is especially cool :-)
Even though the script mines hundreds of blocks very quickly on launch, the _timestamps_ of these blocks are manipulated so they "appear" to be 10 minutes apart. I did this because the pagination API includes parameters for _time_, so I thought it'd be nice to be able to sort the txs by the "virtual" time they were mined.

pinheadmz commented on scripts/pagination-test-wallets.js:
```diff
@@ -0,0 +1,152 @@
+  const feeRate = network.minRelay * 10; // for some reason bc segwit??!!
+  const wdb = wallet.wdb;
+
+  const numInitBlocks = 144 * 3; //. to activate segwit
```
No PR yet, just a head-scratching issue, I could spend more time figuring it out but I don't know if that's a priority since we can just bump the fee:
https://github.com/bcoin-org/bcoin/issues/666

bucko13 replied:
I was talking about the activation rather than the fee: https://github.com/bcoin-org/bcoin/pull/580

pinheadmz replied:
Oh I see! Well there ya go, still a PR. It's ok though just means there's more blocks mined at startup, more funds to disperse... You could bump `numTxBlocks` and `numTxPerBlock` if you want to spread that dough around and mke more txs to paginate...

pinheadmz commented on scripts/pagination-test-wallets.js:
```diff
@@ -0,0 +1,152 @@
+  const mineRegtestBlockToPast = async function(coinbaseAddr) {
+    const entry = await chain.getEntry(node.chain.tip.hash);
+    const job = await miner.createJob(entry, coinbaseAddr);
+    job.attempt.time = virtualNow;
```
yeah! I was playing with this testing actual time-based CSV transactions for the atomic swaps app/guide. Wrote a PR to test CSV-by-time for bcoin:
https://github.com/bcoin-org/bcoin/pull/613

pinheadmz commented on etc/regtest.bcoin.env:
```diff
@@ -12,4 +12,4 @@ BCOIN_NO_WALLET=true
 BCOIN_INDEX_ADDRESS=true
 BCOIN_NODE_HOST=0.0.0.0
-BCOIN_INIT_SCRIPT=funded-dummy-wallets.js
+BCOIN_INIT_SCRIPT=pagination-test-wallets.js
```
I can do that - I guess the Q for you is how do you want to deal with this branch -- because it also targets the bcoin docker container with the pagination repo...

opts:
1) Leave as is, you and mark just grab the whole branch when you test
2) Make this a "real" PR with JUST the script and (like you say) comment out the other launch script (which by the way, still has that `consensus.coinbase_maturity` bug), then it can be merged but won't do any good without the pagination branch of bcoin
3) Make the bcoin docker target the `experimental` branch of bcoin (once that's a real thing) then we can merge, and then ALL bpanel development will be based around that dev branch of bcoin

bucko13 replied:
Option 3! ready to go with bpanel-org/bcoin#experimental :)

pinheadmz replied:
hooray! done with latest commit

pinheadmz commented:
bonus commit: fixing `consensus.COINBASE_MATURITY` bug in `funded-dummy-wallets` script

pinheadmz commented:
This branch is OK to merge now if review is cool. Won't break anything!

bucko13 commented on scripts/pagination-test-wallets.js:
```diff
@@ -0,0 +1,153 @@
+  logger.info('Creating a big mess!...');
+  for (let b = 0; b < numTxBlocks; b++) {
+    for (let t = 0; t < numTxPerBlock; t++) {
+      // TO
```
old comment?

bucko13 replied:
Oh, nvm. I see the `FROM` below. Maybe a little more detail. I thought this was a half written "TODO"

bucko13 commented on scripts/pagination-test-wallets.js:
```diff
@@ -0,0 +1,153 @@
+  await mineRegtestBlockToPast(minerReceive);
+
+  logger.info('Creating a big mess!...');
+  for (let b = 0; b < numTxBlocks; b++) {
```
I'm probably missing something really obvious, but what is `numTxBlocks` for?

pinheadmz replied:
```
  const numInitBlocks = 144 * 3; // blocks mined to primary/default to activate segwit and start chain
  const numTxBlocks = 10; // how many blocks to fill with transactions once wallets are all set up and initial mined funds are distributed
  const numTxPerBlock = 10; // how many txs to generate randomly between accounts before mining a block
  const maxOutputsPerTx = 4; // each random tx has ranomd # of outputs
  const minSend = 50000; // tx amounts are also random
  const maxSend = 100000;
```

bucko13 commented on scripts/pagination-test-wallets.js:
```diff
@@ -0,0 +1,153 @@
+    await node.chain.add(block);
+  };
+
+  const mineRegtestBlockToPast = async function(coinbaseAddr) {
```
love the name :)

bucko13 commented on scripts/pagination-test-wallets.js:
```diff
@@ -0,0 +1,153 @@
+  let virtualNow = network.now() - 60 * 10 * (numInitBlocks + numTxBlocks + 1);
+  const blockInterval = 60 * 10; // ten mimnutes
+
+  const walletNames = [
```
HA! Hilarious.

bucko13 commented:
We can test locally by uncommenting and building locally, but you won’t be able to pull the updated image from docker hub until it gets merged in.

pinheadmz commented:
comments added to explain params

tynes commented:
I really don't like the fact that we need to comment in/comment out parts of the config. Its a major source of problems and is just bad practice. Without some crazy `sed` and/or  `grep` commands, its not possible to automate building the container locally and pulling from dockerhub without inspecting the files and writing custom scripts. Developers shouldn't have to manually edit the file and run a single command, there should just be 2 commands

pinheadmz commented:
Well I think we can only have one `init script` -- maybe just remove `funded-dummy-wallets` altogether? This new script is more verbose anyway, maybe it should just replace?

tynes commented:
![screen shot 2019-01-23 at 9 48 33 am](https://user-images.githubusercontent.com/6626818/51626630-cce62880-1ef4-11e9-8505-bd733af87108.png)

Running into some issues. I commented in the `build` section of the `docker-compose.yml` file for `bcoin` and built it locally, then i ran `docker-compose up bcoin`, watched it generate a ton of transactions, and then it broke

pinheadmz commented:
Huh, are you sure it rebuilt? Might need to wipe out docker container on disk already?
```
 docker stop $(docker ps -a -q); docker rm $(docker ps -a -q); docker volume rm `docker volume ls -q`; docker-compose up bcoin
```

pinheadmz commented:
Addressed your comments: pagination test script is run at `init` and no messages suggesting users comment stuff out. Added a rough try/catch to prevent `wallet already exist` errors. Should be able to `docker-compose up bcoin` multiple times no prob
